k9mail/src/main/java/com/fsck/k9/view/RecipientSelectView.java

```diff
@@ -0,0 +1,271 @@
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.text.TextUtils;
+import android.util.AttributeSet;
+import android.view.LayoutInflater;
+import android.view.View;
+import android.view.inputmethod.InputMethodManager;
+import android.widget.ImageView;
+import android.widget.TextView;
+
+import com.fsck.k9.R;
+import com.fsck.k9.activity.RecipientAdapter;
+import com.fsck.k9.activity.RecipientLoader;
+import com.fsck.k9.view.RecipientSelectView.Recipient;
+import com.fsck.k9.mail.Address;
+import com.tokenautocomplete.TokenCompleteTextView;
+
+
+public class RecipientSelectView extends TokenCompleteTextView<Recipient> implements LoaderCallbacks<List<Recipient>> {
+
+    public static final String ARG_QUERY = "query";
+
+    private RecipientAdapter adapter;
+    private String cryptoProvider;
+    private LoaderManager loaderManager;
+
+    public RecipientSelectView(Context context) {
+        super(context);
+        initView();
+    }
+
+    public RecipientSelectView(Context context, AttributeSet attrs) {
+        super(context, attrs);
+        initView();
+    }
+
+    public RecipientSelectView(Context context, AttributeSet attrs, int defStyle) {
+        super(context, attrs, defStyle);
+        initView();
+    }
+
+    private void initView() {
+        // TODO: validator?
+
+        // don't allow duplicates, based on equality of recipient objects, which is e-mail addresses
+        allowDuplicates(false);
+
+        // if a token is completed, pick an entry based on best guess
+        performBestGuess(true);
+
+        adapter = new RecipientAdapter(getContext());
+        setAdapter(adapter);
+    }
+
+    @Override
+    protected View getViewForObject(Recipient recipient) {
+        LayoutInflater l = LayoutInflater.from(getContext());
+        @SuppressLint("InflateParams") // root will be attached by caller
+        View view = l.inflate(R.layout.recipient_token_item, null, false);
+
+        // A name is preferred, but show if nothing is available
+        TextView vName = (TextView) view.findViewById(android.R.id.text1);
+        String personal = recipient.address.getPersonal();
+        if ( ! TextUtils.isEmpty(personal)) {
+            vName.setText(personal);
+        } else {
+            vName.setText(recipient.address.getAddress());
+        }
+
+        ImageView contactView = (ImageView) view.findViewById(R.id.contact_photo);
+        RecipientAdapter.setContactPhotoOrPlaceholder(getContext(), contactView, recipient);
+
+        View cryptoStatusRed = view.findViewById(R.id.contact_crypto_status_red);
+        View cryptoStatusOrange = view.findViewById(R.id.contact_crypto_status_orange);
+        View cryptoStatusGreen = view.findViewById(R.id.contact_crypto_status_green);
+
+        boolean hasCryptoProvider = cryptoProvider != null;
+        // display unavailable status only if a crypto provider is even available
+        cryptoStatusRed.setVisibility(
```

Comment:
The code is easier to read if you make the conditions first-level elements, i.e.
```java
if (!hasCryptoProvider) {
    cryptoStatusRed.setVisibility(View.GONE);
    cryptoStatusOrange.setVisibility(View.GONE);
    cryptoStatusGreen.setVisibility(View.GONE);
} else if (recipient.cryptoStatus == RecipientCryptoStatus.UNAVAILABLE) {
    cryptoStatusRed.setVisibility(View.VISIBLE);
    cryptoStatusOrange.setVisibility(View.GONE);
    cryptoStatusGreen.setVisibility(View.GONE);
} else if (recipient.cryptoStatus == RecipientCryptoStatus.AVAILABLE_UNTRUSTED) {
    cryptoStatusRed.setVisibility(View.GONE);
    cryptoStatusOrange.setVisibility(View.VISIBLE);
    cryptoStatusGreen.setVisibility(View.GONE);
} else if (recipient.cryptoStatus == RecipientCryptoStatus.AVAILABLE_TRUSTED) {
    cryptoStatusRed.setVisibility(View.GONE);
    cryptoStatusOrange.setVisibility(View.GONE);
    cryptoStatusGreen.setVisibility(View.VISIBLE);
}
```